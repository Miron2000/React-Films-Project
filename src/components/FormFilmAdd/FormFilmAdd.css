.page__add-film {
    margin-top: 2rem;
    margin-bottom: 5rem;
}

.title__add-film {
    font-family: "Noto Sans CJK SC";
    font-size: 2rem;
    text-align: center;
    color: #86b5c2;
}

.container__form {
    display: flex;
    justify-content: center;
}

.form__add-film {
    width: 590px;
    padding: 20px 30px;
    background: #86b5c2;
    border: 2px solid #b539c4;
    border-radius: 8px;
    /*position: relative;*/
}

.container__add-film {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
}

.fields__first {
    display: flex;
    position: relative;
}

.fields__second {
    display: flex;
    margin-top: 1rem;
    position: relative;
}

.fields__third {
    margin-top: 1rem;
}

.fields__fourth {
    display: flex;
    position: relative;
}

.star__required-field{
    color: red;
    position: absolute;
    z-index: 1;
    right: 14rem;
}

.star__required-field_fields__fourth {
    color: red;
    position: absolute;
    z-index: 1;
    top: 18px;
    right: 132px;
}

.MuiInputLabel-shrink {
    margin-top: -7px !important;
    color: white !important;
}

.input__text {
    margin-left: 25px !important;
}

.btn__indent {
    margin-left: 16px !important;
}

.dropdown-heading {
    height: 56px !important;
}

.input__name {
    width: 371px;
    background-color: ghostwhite;
    border-radius: 5px;
}

.input__releaseDate {
    width: 195px;
    background-color: ghostwhite;
    border-radius: 5px;
}

.input__assessment {
    width: 82px;
    background-color: ghostwhite;
    border-radius: 5px;
    height: 56px;
}

.input__imdb {
    width: 88px;
    background-color: ghostwhite;
    border-radius: 5px;
    height: 56px;
}

.input__overview {
    width: 592px;
    background-color: ghostwhite;
    border-radius: 5px;
}

.btn__genres {
    background-color: #7ae387 !important;
    margin-top: 1rem !important;
    border: 1px solid #b539c4 !important;
    cursor: pointer !important;
    height: 55px !important;
}

.btn__submit:disabled {
    background-color: #ada1a6eb !important;
    margin-top: 1rem !important;
    border: 1px solid #ada1a6eb !important;
    cursor: no-drop !important;
}

.btn__submit:enabled {
    background-color: #b539c4 !important;
    margin-top: 1rem !important;
    border: 1px solid #b539c4 !important;
    cursor: pointer !important;
}

@media (max-width: 768px) {
    .form__add-film {
        width: auto;
        margin: auto;
    }

    .container__form {
        flex-direction: column;
    }

    .fields__first {
        flex-direction: column;
    }

    .fields__second {
        flex-direction: column;
    }
    .star__required-field {
        right: 0;
    }

    .input__name {
        width: 245px;
    }

    .input__releaseDate {
        width: 245px;
    }

    .input__assessment {
        width: 245px;
    }

    .input__imdb {
        width: 245px;
    }

    .input__overview {
        width: 245px;
    }

    .input__text {
        margin-top: 20px !important;
        margin-left: 0 !important;
    }

}