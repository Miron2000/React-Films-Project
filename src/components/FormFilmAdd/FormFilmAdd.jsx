import React, {useEffect, useState} from 'react';
import './FormFilmAdd.css';
import TextField from "@material-ui/core/TextField";
import Button from '@material-ui/core/Button';
import Icon from '@material-ui/core/Icon';
import MenuItem from '@material-ui/core/MenuItem';
import InputGenres from "./InputGenres/InputGenres";
import InputCountries from "./InputCountries/InputCountries";
import preloader from "../../preloader/Eclipse-1s-200px.gif";

function FormFilmAdd(props) {

    const [inputFields, setInputFields] = useState(
        {name: '', genres: [''], releaseDate: '', countries: [], assessment: '', imdbFilm: '', overview: ''}
    )
    const [isLoading, setIsLoading] = useState(false);
    const [disabled, setDisabled] = useState(true);

    const handleChangeInput = (event) => {
        const values = {...inputFields};
        values[event.target.name] = event.target.value;
        setInputFields(values)
    }

    const onFieldChangeCountries = (fieldName, value) => {
        const values = {...inputFields};
        values[fieldName] = value;
        setInputFields(values);
        // console.log(values[fieldName], 'values[fieldName]');

        // if( values[fieldName].length > 0) {
        //     setDisabled(false);
        // }  else {
        //     setDisabled(true);
        // }
    }

    const onFieldChangeGenres = (fieldName, value, index) => {
        console.log(fieldName, 'fieldName');
        console.log(value, 'value')
        const values = {...inputFields};
        values[fieldName] = value;
        setInputFields(values);
        // console.log(values[fieldName], 'values[fieldName]');

        // if( values[fieldName].length > 0) {
        //     setDisabled(false);
        // }  else {
        //     setDisabled(true);
        // }
    }

    // useEffect(() => {
    //     if(inputFields.genres.length > 0 && inputFields.countries.length > 0 && inputFields.name.length > 0) {
    //         setDisabled(false);
    //     } else {
    //         setDisabled(true);
    //     }
    // }, [inputFields.genres, inputFields.countries])

    const nameHandler = (fieldName, event) => {
        const values = {...inputFields};
        values[fieldName] = event.target.value;
        setInputFields(values);

        if (values[fieldName].length > 0) {
            setDisabled(false);
        } else {
            setDisabled(true);
        }

    }

    const addSelectGenre = () => {
        setInputFields({...inputFields, genres: [...inputFields.genres, '']})
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        console.log(inputFields, 'inputFields')

        fetch('http://localhost:3000/api/film', {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-type': 'application/json'
            },
            body: JSON.stringify({
                name: inputFields.name,
                releaseDate: inputFields.releaseDate.length > 0 ? inputFields.releaseDate.split("-").reverse().join('.') : '-',
                assessment: inputFields.assessment.toString().length > 0 ? inputFields.assessment : '-',
                imdbFilm: inputFields.imdbFilm.length > 0 ? inputFields.imdbFilm : '-',
                // country_id: inputFields.countries[0].value,
                // genre_id: inputFields.genres.target.value,
                country_id: inputFields.countries,
                genre_id: inputFields.genres,
                overview: inputFields.overview.length > 0 ? inputFields.overview : '-'
            })
        })
            .then(res => {
                res.json()
            }).catch(err => {
            console.log(err)
        })

        setInputFields({
            name: '',
            genres: [],
            releaseDate: '',
            countries: [],
            assessment: '',
            imdbFilm: '',
            overview: ''
        });
    }

    return (
        <div className='page__add-film'>
            {isLoading ? <img className='preloader' src={preloader}/> : null}
            <h1 className='title__add-film'>Add new Film</h1>
            <div className="container__form" style={isLoading ? {display: 'none'} : {display: ''}}>
                <form className='form__add-film' onSubmit={handleSubmit}>
                    <div className='container__add-film'>

                        <div className='fields__first'>
                            <TextField
                                className="input__name"
                                name="name"
                                label="Film Name"
                                value={inputFields.name}
                                variant="outlined"
                                onChange={(event) => nameHandler('name', event)}

                            />

                            <TextField
                                className='input__text input__releaseDate'
                                type="date"
                                name="releaseDate"
                                label="Release date"
                                variant="outlined"
                                value={inputFields.releaseDate}
                                onChange={event => handleChangeInput(event)}
                                InputLabelProps={{
                                    shrink: true,
                                }}
                            />
                        </div>

                        <div className='fields__second'>

                            <InputCountries onChange={(value) => onFieldChangeCountries('countries', value)}
                                            value={inputFields.countries} isLoading={isLoading}
                                            setIsLoading={setIsLoading}/>

                            <TextField
                                className='input__text input__assessment'
                                name="assessment"
                                label="Rating"
                                type="number"
                                value={inputFields.assessment}
                                InputProps={{
                                    inputProps: {
                                        max: 10, min: 1
                                    }
                                }}
                                variant="outlined"
                                onChange={event => handleChangeInput(event)}

                            />

                            <TextField
                                className='input__text input__imdb'
                                id="select"
                                label="IMDB"
                                name="imdbFilm"
                                select
                                variant="outlined"
                                onChange={event => handleChangeInput(event)}>
                                <MenuItem value="Yes">Yes</MenuItem>
                                <MenuItem value="No">No</MenuItem>
                            </TextField>
                        </div>
                        <div className='fields__third'>
                            <TextField
                                className="input__overview"
                                name="overview"
                                rows={10}
                                label="Overview"
                                value={inputFields.overview}
                                variant="outlined"
                                onChange={event => handleChangeInput(event)}
                            />
                        </div>
                        <div className='fields__fourth'>
                            <div>
                                {inputFields.genres.map((genre, index) => {
                                    console.log(genre, 'genre')
                                    return (
                                        <InputGenres key={index}
                                                     onChange={(event) => onFieldChangeGenres('genres', event.target.value, index)}
                                                     value={inputFields.genres}
                                                     isLoading={isLoading}
                                                     setIsLoading={setIsLoading}/>)
                                })}
                            </div>
                            <Button
                                className="btn__genres btn__indent"
                                variant="contained"
                                onClick={addSelectGenre}>Add Genre
                            </Button>
                        </div>
                    </div>

                    <Button
                        className="btn__submit"
                        variant="contained"
                        type="submit"
                        disabled={disabled}
                        endIcon={<Icon>send</Icon>}
                        onClick={handleSubmit}>Add Film
                    </Button>

                </form>
            </div>
        </div>
    );
}

export default FormFilmAdd;