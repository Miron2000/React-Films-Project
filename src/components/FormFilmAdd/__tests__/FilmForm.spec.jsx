import React from "react";
import {shallow, mount, render} from '../../../enzyme';
import FormFilmAdd from "../FormFilmAdd";

jest.mock('reactjs-alert', () => ({
    __esModule: true,
    default: () => 'ReactJsAlert',
}));

describe.only("should render FormFilmAdd component", () => {
    it('should contain .form__add-film wrapper', () => {
        const component = shallow(<FormFilmAdd/>);
        console.log(component.debug())

        const wrapper = component.find('.form__add-film');
        console.log(wrapper.debug())
        expect(wrapper.length).toBe(1);
    });

    it('should contain textfield', () => {
        const component = shallow(<FormFilmAdd/>);
        const wrapper = component.find('input');
        expect(wrapper.length).toBe(0);
    });
})

describe("FormFilmAdd component", () => {
    it('should render FormFilmAdd component', () => {
        const component = shallow(<FormFilmAdd/>);
        expect(component).toMatchSnapshot();
    });

    it('should renders <InputCountries/> component', () => {
        const component = shallow(<FormFilmAdd/>);
        console.log(component.debug())
        expect(component.find('InputCountries')).toHaveLength(1);
    });
})

