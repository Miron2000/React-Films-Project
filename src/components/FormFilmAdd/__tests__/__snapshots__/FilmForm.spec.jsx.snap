// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFilmAdd component should render FormFilmAdd component 1`] = `ShallowWrapper {}`;
